File: haskell/fibo.hs




fib :: Integer -> Integer
fib 0 = 0
fib 1 = 1
fib n = (fib (n-1)) + (fib (n-2))

fibTuple :: (Integer,Integer,Integer) -> (Integer,Integer,Integer)
-- fibTuple (fibN,fib(N+1),N)
fibTuple (a,b,0) = (a,b,0)
fibTuple (a,b,n) = (d,(c+d),n)
	 where (c,d,e) = fibTuple (a,b,(n-1))

fastFib :: Integer -> Integer
fastFib n = getFirst (fibTuple (0,1,n))

getFirst :: (a,a,a) -> a
getFirst (x,_,_) = x

File: prolog/fibo.pl


fib(0,0).
fib(1,1).
fib(N,FibN):-
    length(_,N),
    N > 1,
    NM1 is N - 1,
    NM2 is N - 2,
    fib(NM1,FibNM1),
    fib(NM2,FibNM2),
    FibN is FibNM1 + FibNM2.

fibTuple((0,1,0)).
fibTuple((FibN,FibNP1,N)):-
    length(_,N),
    N > 0,
    NM1 is N - 1,
    fibTuple((FibNM1,FibN,NM1)),
    FibNP1 is FibNM1 + FibN.

fastFib(N,FibN):-
    fibTuple((FibN,_,N)).
